neevaco/neeva-android
# Show warning when screen overlay is detected

If we detect that another app is passing motion events through itself to us via the "Draw over other apps" permission, show the user a one-time dialog that tells them that they might be doing something dangerous and allow them the opportunity to take the permission away from the other app.  This only matters on Android versions < 12 because they mitigated it at the system level >= Android 12.

* Add a `ScreenOverlayDetectingFrameLayout` that wraps all of our Views and tries to detect when an overlay is passing `MotionEvent`s through itself.  If it detects an obscured touch, it will throw up a dialog telling them.  We only look for the OBSCURED flag instead of PARTIALLY_OBSCURED because it would negatively affect usage of the app.

* Add a `OverlayWarningDialogManager` that encapsulates the logic for detecting the overlays.

* Adds a 5 minute timeout to the instrumentation tests to try and catch when CircleCI fails to keep running our tests.

This currently doesn't do anything during First Run; we'll probably want to add support for that once the user can actually sign up for an account there.

## Checklists

### Pointers to help with review
* Figma: https://www.figma.com/file/D3pbLY6ZP5YtCjTYqsGOKO/Android-Design-System?node-id=2331%3A38054
* Tapjacking: https://developer.android.com/topic/security/risks/tapjacking

### How was this tested?

### Manual test cases
Base test steps:
1. Install another app (like Twilight) that draws over other apps
2. Start the other app
3. Start our app
4. Click anywhere on the screen
5. See the dialog pop up

Dismiss test:
6. Click on the "Don't show this again" button
7. See the dialog go away
8. Click anywhere again
9. Confirm no dialog appears

Settings test:
6. Click on the "Open Settings" button
7. Get sent to Android Settings
8. Come back to the app
9. See the dialog is gone
10. Click anywhere again
11. Confirm no dialog appears

Back button test:
6. Hit system back
7. See that the dialog stays on the screen
8. Click outside of the dialog on the scrim
9. See that the dialog stays on the screen
10. Click on the "Don't show this again" button
11. See the dialog go away

### Screenshots and screen recordings

<img src=https://user-images.githubusercontent.com/96092902/197898385-a1d7db59-de17-4eec-9c78-11028f60a453.png width=500>

Clicking outside of the scrim does nothing and you can only dismiss by clicking one of the two dialog buttons:

https://user-images.githubusercontent.com/96092902/197897597-95fcd240-ea5b-4893-93f9-59070148cd97.mp4

Going to settings and coming back doesn't show the dialog again:

https://user-images.githubusercontent.com/96092902/197897603-35d0aa8c-f387-4458-b0cb-a944fb665a97.mp4
